cmd/thanos/compact.go

```diff
@@ -420,6 +420,15 @@ func runCompact(
 	}
 
 	compactMainFn := func() error {
+		// TODO(bwplotka): Find a way to avoid syncing if no op was done.
+		if err := sy.SyncMetas(ctx); err != nil {
+			return errors.Wrap(err, "sync before retention")
+		}
+
+		if err := compact.ApplyRetentionPolicyByResolution(ctx, logger, bkt, sy.Metas(), retentionByResolution, compactMetrics.blocksMarked.WithLabelValues(metadata.DeletionMarkFilename, "")); err != nil {
```

Comment:
One concern I have is that some raw blocks might beyond the retention. But they are able to be downsampled to 5m blocks that still within the 5m block retention period.
Changing the order makes users configure a longer retention for 5m blocks to get those downsampled.

Reply:
@yeya24 Yeah, this makes sense. So basically for this to be safe, `ApplyRetentionPolicyByResolution` would need to also check if we have downsampled a block already. And only apply retention if it's been downsampled.

However, this defeats the point (with the current implementation), as downsampling only occurs after compaction. So we are where we started again, unless we also perform downsampling before compaction, and have a check in the downsampler that ensures we are at the right compaction level for downsampling.

Reply:
Moving retention to the beginning could also mistakenly deleted blocks that are expected to be compacted. Since currently retention is calculated using the block's max time, so after a compaction we may keep the block instead of deleting all these small blocks.
Corretness wise I don't think this is safe to do.

Reply:
I wonder if we could handle this case by simulating all compaction + downsampling operations, and if the resulting block would be outside of retention, mark it for deletion immediately.

The specific case that this would help with: Lowering retention to purge the backlog. It would come at the cost of additional complexity though, so I'm not sure it's worth it.

Reply:
If it is for purging the backlog we have the bucket tool retention already. Does that satisfy your usecase?

Reply:
Ahh, TIL about `bucket tool retention`. Though I suspect it suffers from the same issue we discussed here: it does not take into account downsampling, so we run the risk of purging blocks that have not yet been downsampled.

In other words, using that tool is unsafe for the same reason that this proposed patch is unsafe.

Reply:
Right, it is the same as this patch so you use it at your own risk.
If that's the same then do we still need this patch? If you have other things to improve we can discuss on the issue https://github.com/thanos-io/thanos/issues/1708

Reply:
Agreed, I think we can close this. 👍